Add HCP TF metrics service usage documentation [IPE-1493] (#2480)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/users-teams-organizations/organizations/metrics.mdx

---
page_title: Collect organization metrics
description: Learn how to create a metrics service token and collect metrics from your HCP Terraform organization.
tfc_only: true
---

# Collect organization metrics

HCP Terraform provides an API endpoint that you can use to gather metrics related to your HCP Terraform organization.

## Introduction

When you create a metrics service token, HCP Terraform automatically begins gathering metrics from your organization and makes them available in a Prometheus-compatible format. If your organization has no valid metrics service tokens, HCP Terraform stops gathering metrics until you create a new valid token.

### Workflow

To gather metrics from your HCP Terraform organization, complete the following steps:

1. Create a metrics service API token. Refer to [Create a metrics service token](#create-a-metrics-service-token) for more information.
1. Configure your metrics collection software to gather metrics from HCP Terraform. Refer to [Gather metrics](#gather-metrics) for more information.

## Requirements

> To manage your organization's metrics service API token, you must be a member of your organization's "owners" team. Refer to the [Organization owners](/terraform/cloud-docs/users-teams-organizations/permissions/organization#organization-owners) documentation for more information on the owners team.

Reviewer: Is there a minimum HCP TF plan needed for metrics?

> To manage your organization's metrics service API token, you must be a member of your organization's "owners" team. Refer to the [Organization owners](/terraform/cloud-docs/users-teams-organizations/permissions/organization#organization-owners) documentation for more information on the owners team.

Reviewer 2: ```suggestion
To manage your organization's metrics service API token, you must be a member of your organization's **owners** team. Refer to the [Organization owners](/terraform/cloud-docs/users-teams-organizations/permissions/organization#organization-owners) documentation for more information on the owners team.
```
Because it's a UI label

## Create a metrics service token

To gather metrics from HCP Terraform, you must provide a metrics service token for your organization. Complete the follow steps to generate a metrics service API token:

1. Sign in to [HCP Terraform](https://app.terraform.io/) and navigate to your organization.
1. Choose **Settings** from the sidebar, then **API tokens**.
1. Choose the **Metrics Service Tokens** tab.
1. Click **Generate a token**.
1. Give the token a **Name** and **Expiration**. 
1. Click **Generate**.

> Save this token in a safe place since HCP Terraform only displays it once. You will use it to authenticate your metrics collection software.

Reviewer: Using "because" instead of "since" as a conjunction is often clearer.  Less need to mentally backtrack if it's interpreted as a time reference.

(Very much a nitpick)

> Save this token in a safe place since HCP Terraform only displays it once. You will use it to authenticate your metrics collection software.

Reviewer 2: ```suggestion
Save this token in a safe place since HCP Terraform only displays it once. You must present the token to authenticate your metrics collection software.
```

You can also manage your metrics service API tokens with the HCP Terraform API. Refer to the [Metrics service tokens API reference](/terraform/cloud-docs/api-docs/metrics-service-tokens) for more information.

## Gather metrics

To read metrics from HCP Terraform, configure your metrics collection software to send a `GET` request with your organization ID and metrics service token to the `/metrics` endpoint.

`GET https://api.cloud.hashicorp.com/telemetry/hcpt/$ORG_ID/metrics`

| Header key | Value |
| ---------- | ----- |
| `Authorization` | `Bearer $TOKEN` |

> The following is example Prometheus configuration to gather metrics from HCP Terraform.

Reviewer 2: ````suggestion
To read metrics from HCP Terraform, configure your metrics collection software to send a `GET` request with your organization ID and metrics service token to the [`/metrics`]() endpoint. You must present the metrics service token you generated in [Generate metrics service token](#generate-metrics-service-token) to call the endpoint.

For example:

```shell-session
$ curl \
    --header "Authorization: Bearer $TOKEN \
    --request GET
    https://api.cloud.hashicorp.com/telemetry/hcpt/$ORG_ID/metrics
```  

The following is example Prometheus configuration to gather metrics from HCP Terraform.
````
I couldn't find the metrics endpoint docs to set as the link. Is that coming later? 

The tabular presentation of the authorization key is a little jarring because that is usually a reference page format.

<CodeBlockConfig filename="prometheus.yml">

```yaml
global:
  scrape_interval: 15s

scrape_configs:
  - job_name: 'hcp_telemetry'
    scheme: https
    metrics_path: '/telemetry/hcpt/org-XXXXXXXXXXXXX/metrics'

    authorization:
      type: 'Bearer'
      credentials: 'hcpt_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'

    static_configs:
      - targets:
        - 'api.cloud.hashicorp.com:443
```

</CodeBlockConfig>

> ## Metrics reference

Reviewer 2: I think we should move these to a separate topic under a /reference folder, similar to the [Nomad metrics reference docs](https://developer.hashicorp.com/nomad/docs/reference/metrics) and link to them from this page.  In the navigation, it would live next the API and System architecture reference pages

HCP Terraform collects the following metrics:

| Metric | Labels | Type | Description | Update Interval |
| ------ | ------ | ---- | ----------- | --------------- |
| `tfc_vcs_requests_total` | organization_id, workspace, project, status, vcs.provider | count | Total number of requests from HCP Terraform to VCS providers | 60s | 

> | `tfc_vcs_webhook_exists_checks_total` | organization_id, workspace, project, status, vcs.provider | count | Total webhook existence checks from HCP Terraform to VCS providers | 60s | 

Reviewer 2: ```suggestion
| `tfc_vcs_webhook_exists_checks_total` | organization_id, workspace, project, status, vcs.provider | count | Total number of checks that verify existing webhooks from HCP Terraform to VCS providers | 60s | 
```
Right? This is what we mean by "existence check"?

| `vcs_tfc_requests_total` | organization_id, workspace, project, status, vcs.provider | count | Webhook requests from VCS providers to HCP Terraform | 60s | 
| `tfc_vcs_requests_duration_seconds` | organization_id, workspace, project, vcs.provider | histogram | Duration of HCP Terraform-to-VCS API requests in seconds | 60s | 

> | `tfc_vcs_webhook_exists_checks_duration_seconds` | organization_id, project, workspace, vcs.provider | histogram | Duration of webhook existence checks in seconds | 60s | 

Reviewer: Complete nitpick, but this is the only row where the order is `project, workspace` instead of `workspace, project`.

> | `tfc_vcs_webhook_exists_checks_duration_seconds` | organization_id, project, workspace, vcs.provider | histogram | Duration of webhook existence checks in seconds | 60s | 

Reviewer 2: ```suggestion
| `tfc_vcs_webhook_exists_checks_duration_seconds` | organization_id, project, workspace, vcs.provider | histogram | Duration in seconds of checks that verify existing webhooks  | 60s | 
```
Is it an average of each check? Is there a separate entry for each check?

> | `tfc_terraform_agent_count` | organization_id, status (idle/busy/unknown/errored/exited), agent_pool_id | gauge | Number of agents in different status by pool. | 60s | 

Reviewer 2: ```suggestion
| `tfc_terraform_agent_count` | organization_id, status (idle/busy/unknown/errored/exited), agent_pool_id | gauge | Number of agents in any status by pool. | 60s | 
```
Or do we really mean that we're only counting agents that have different statuses? If so, what is the baseline status? For example, does the metric return `3` if one agent is active and three agents are pending or some other state?

> | `terraform_run_phase_transitions_total` | organization_id, workspace_id, phase | count | Cumulative count of run transitions into each Terraform execution phase | 60s | 

Reviewer 2: Is this description clear to the target audience? I don't understand what we mean.  Are we capturing the number of run phase transitions during the 60s interval? I understood the "Interval update" column to be when the platform takes a snapshot of the data. If it's a static snapshot, how can it register a dynamic event?

> | `terraform_run_status_total` | organization_id, workspace_id, run_status | count | Cumulative count of completed runs by final status | 60s | 

Reviewer 2: For all intents and purposes, is the basically a count of finished and errored runs?
